CVS-51750 Model zoo custom node use case documentation (#622)
openvinotoolkit/model_server: docs/faces_analysis_dag.md

# Usage of Object Detection and Recognition models from OpenVINO Model Zoo in DAG

## Analysis of multiple faces in single image frame request
This document demonstrates how to create complex pipelines using object detection and object recognition models from OpenVINO Model Zoo. As an example, we will use [face-detection-retail-0004](https://github.com/openvinotoolkit/open_model_zoo/blob/master/models/intel/face-detection-retail-0004/description/face-detection-retail-0004.md) to detect multiple faces on the image. Then, for each detected face we will crop it using [model_zoo_intel_object_detection](../src/custom_nodes/model_zoo_intel_object_detection) example custom node. Finally, each image face image will be forwarded to [age-gender-recognition-retail-0013](https://github.com/openvinotoolkit/open_model_zoo/blob/master/models/intel/age-gender-recognition-retail-0013/description/age-gender-recognition-retail-0013.md) and [emotion-recognition-retail-0003](https://github.com/openvinotoolkit/open_model_zoo/blob/master/models/intel/emotions-recognition-retail-0003/description/emotions-recognition-retail-0003.md) models.

> ![OCR graph](faces_analysis.png)

Reviewer: this is not OCR

Author: done

Using such pipeline, a single request to OVMS can perform a complex set of operations to determine all faces and its properties.

## Pipeline Configuration Graph

Below is depicted graph implementing faces analysis pipeline execution. 

![Faces Analysis Pipeline Graph](faces_analysis_graph.svg)

It includes the following Nodes:

> - Model `face-detection` - inference execution which takes user image as input. It returns outputs including information about face coordinates and confidence levels.

Reviewer 2: ```suggestion
- Model `face-detection` - deep learning model which takes user image as input. Its outputs contain information about face coordinates and confidence levels.
```

Author: Done

### Face detection model
```
wget https://storage.openvinotoolkit.org/repositories/open_model_zoo/2021.3/models_bin/2/age-gender-recognition-retail-0013/FP32/age-gender-recognition-retail-0013.xml
wget https://storage.openvinotoolkit.org/repositories/open_model_zoo/2021.3/models_bin/2/age-gender-recognition-retail-0013/FP32/age-gender-recognition-retail-0013.bin
```
### Age gender recogonition model
```
wget https://storage.openvinotoolkit.org/repositories/open_model_zoo/2021.3/models_bin/2/face-detection-retail-0004/FP32/face-detection-retail-0004.xml
wget https://storage.openvinotoolkit.org/repositories/open_model_zoo/2021.3/models_bin/2/face-detection-retail-0004/FP32/face-detection-retail-0004.bin
```

### Emotions recognition model
```
wget https://storage.openvinotoolkit.org/repositories/open_model_zoo/2021.3/models_bin/2/emotions-recognition-retail-0003/FP32/emotions-recognition-retail-0003.xml
wget https://storage.openvinotoolkit.org/repositories/open_model_zoo/2021.3/models_bin/2/emotions-recognition-retail-0003/FP32/emotions-recognition-retail-0003.bin
```

## Building the Custom Node "model_zoo_intel_object_detection" Library 

> Custom nodes are loaded into OVMS as dynamic library implementing OVMS API from [custom_node_interface.h](../src/custom_node_interface.h).

Reviewer 2: ```suggestion
Custom nodes are loaded into OVMS as dynamic libraries implementing OVMS API from [custom_node_interface.h](../src/custom_node_interface.h).
```

Author: Done

- go to the custom node source code folder [src/custom_nodes/model_zoo_intel_object_detection](../src/custom_nodes/model_zoo_intel_object_detection)
- run `make` command

This command will export the compiled library in `./lib` folder.
Copy this `lib` folder to the same location with previously downloaded models.

## OVMS Configuration File

The configuration file for running the OCR demo is stored in [config.json](../src/custom_nodes/model_zoo_intel_object_detection/config.json).
Copy this file along with the model files.

## Final directory structure
```
workspace
├── age-gender-recognition-retail-0013
│   └── 1
│       ├── age-gender-recognition-retail-0013.bin
│       └── age-gender-recognition-retail-0013.xml
├── config.json
├── emotion-recognition-retail-0003
│   └── 1
│       ├── emotions-recognition-retail-0003.bin
│       └── emotions-recognition-retail-0003.xml
├── face-detection-retail-0004
│   └── 1
│       ├── face-detection-retail-0004.bin
│       └── face-detection-retail-0004.xml
└── lib
    └── libcustom_node_model_zoo_intel_object_detection.so
```

## Deploying OVMS

> Deploy OVMS with OCR demo pipeline using the following command:

Reviewer: This is not OCR

Author: done
